### PR TITLE
Migrate tools/licenses to null safety

### PR DESCRIPTION
**Pull request by mit-mit**
Migrate `src/flutter/tools/licenses` to null safety.

Contributes to https://github.com/flutter/flutter/issues/110020

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by zanderso** [comment]
From PR triage: Adding the WIP label.

**Comment by mit-mit** [comment]
I believe this is ready for a review.

There is [one test failure](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8803233766888408337/+/u/licenses_check/stdout?format=raw), but I'm a bit puzzled as I don't think I changed anything in the logic of [`_findLicenseBlocks`](https://github.com/flutter/engine/pull/35882/files#diff-609f3666ecaf62f888d104d1475eb98c44b64fb10acc80555ab1919dffa1a4dbR712). 

And the test output actually looks legit; it's triggered by the file `third_party/libjpeg-turbo/simd/jsimd_mips_dspr2.S` that has several copyright statements... @Hixie @goderbauer looks you both worked on this tool, so wondering if you can comment if that failure is actually valid?

File: tools/licenses/lib/main.dart [hunk 1]
```diff
@@ -39,10 +39,10 @@ abstract class _RepositoryEntry implements Comparable<_RepositoryEntry> {
 abstract class _RepositoryFile extends _RepositoryEntry {
   _RepositoryFile(_RepositoryDirectory parent, fs.File io) : super(parent, io);
 
-  Iterable<License> get licenses;
+  Iterable<License>? get licenses;
 
   @override
-  String get libraryName => parent.libraryName;
+  String get libraryName => parent?.libraryName ?? '';
```
Comment (Hixie) [inline tools/licenses/lib/main.dart:45]: why `?? ''`?

Comment (Hixie) [reply tools/licenses/lib/main.dart:45]: i think this should be `parent!`

File: tools/licenses/lib/licenses.dart [hunk 2]
```diff
@@ -923,22 +921,22 @@ Iterable<_LicenseMatch> _tryReferenceByType(String body, RegExp pattern, License
   }
 }
 
-License _dereferenceLicense(int groupIndex, String Function(int index) group, MultipleVersionedLicenseReferencePattern pattern, LicenseSource parentDirectory, { String origin }) {
-  License result = pattern.checkLocalFirst ? parentDirectory.nearestLicenseWithName(group(groupIndex)) : null;
+License _dereferenceLicense(int groupIndex, String? Function(int? index) group, MultipleVersionedLicenseReferencePattern pattern, LicenseSource parentDirectory, { String? origin }) {
```
Comment (Hixie) [inline tools/licenses/lib/licenses.dart:924]: i think `String? Function(int index) group` would be closer to the original intent. requires some `!`s below but removes some ugly casts later.

File: tools/licenses/lib/main.dart [hunk 3]
```diff
@@ -2793,19 +2799,20 @@ void _writeSignature(String signature, system.IOSink sink) {
 // Checks for changes to the license tool itself.
 //
 // Returns true if changes are detected.
-Future<bool> _computeLicenseToolChanges(_RepositoryDirectory root, {String goldenSignaturePath, String outputSignaturePath}) async {
+Future<bool> _computeLicenseToolChanges(_RepositoryDirectory root, { required String goldenSignaturePath,required String outputSignaturePath }) async {
```
Comment (Hixie) [inline tools/licenses/lib/main.dart:2802]: trivial nit: missing space after comma

**Comment by Hixie** [comment]
> There is [one test failure](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8803233766888408337/+/u/licenses_check/stdout?format=raw), but I'm a bit puzzled as I don't think I changed anything in the logic of [`_findLicenseBlocks`](https://github.com/flutter/engine/pull/35882/files#diff-609f3666ecaf62f888d104d1475eb98c44b64fb10acc80555ab1919dffa1a4dbR712).

Looking at the goldens, it seems we used to do this one correctly, so something in this patch must have affected it...

File: tools/licenses/lib/main.dart [hunk 4]
```diff
@@ -462,40 +461,40 @@ class _RepositoryLibJpegTurboLicense extends _RepositoryLicenseFile {
...
+  List<License>? _licenses;
 
   @override
-  List<License> get licenses {
-    if (_licenses == null) {
-      final _RepositoryReadmeIjgFile readme = parent.getChildByName('README.ijg') as _RepositoryReadmeIjgFile;
-      final _RepositorySourceFile main = parent.getChildByName('turbojpeg.c') as _RepositorySourceFile;
-      final _RepositoryDirectory simd = parent.getChildByName('simd') as _RepositoryDirectory;
+  List<License>? get licenses {
+    if (_licenses == null && parent != null) {
```
Comment (Hixie) [inline tools/licenses/lib/main.dart:468]: the `&& parent != null` should not be necessary

File: tools/licenses/lib/licenses.dart [hunk 5]
```diff
@@ -650,14 +647,15 @@ _SplitLicense _splitLicense(String body, { bool verifyResults = true }) {
...
+      bool hadMoreLines;
+      while (hadMoreLines = lines.moveNext() && lines.current.value.startsWith(prefix)) {
         final String nextAuthor = lines.current.value.substring(prefix.length);
         if (nextAuthor == '' || nextAuthor[0] == ' ' || nextAuthor[0] == '\t') {
           throw 'unexpectedly ragged author list when looking for copyright';
         }
         end = lines.current.end;
       }
-      if (lines.current == null) {
+      if (!hadMoreLines) {
```
Comment (Hixie) [inline tools/licenses/lib/licenses.dart:658]: i wonder if this is somehow not 100% exactly semantically equivalent and is causing the bug you ran into

File: tools/licenses/lib/main.dart [hunk 6]
```diff
@@ -546,37 +545,42 @@ class _RepositoryFreetypeLicenseFile extends _RepositoryLicenseFile {
...
+  List<License>? _targetLicense;
 
   void _warmCache() {
-    _targetLicense ??= <License>[parent.nearestLicenseWithName(_target)];
+    if (parent != null && _targetLicense == null) {
+      final License? license = parent!.nearestLicenseWithName(_target);
+      if (license != null) {
+        _targetLicense = <License>[license];
+      }
+    }
```
[the remark this answers is not in the document]
Comment (mit-mit) [reply tools/licenses/lib/main.dart:565]: Done

File: tools/licenses/lib/main.dart (hunk 4 above)
Comment (mit-mit) [reply tools/licenses/lib/main.dart:468]: Done.  I'm a bit puzzled why that works given `parent` is nullable. I guess there is an invariant that `get` isn't called when `parent` is null?

File: tools/licenses/lib/licenses.dart (hunk 2 above)
Comment (mit-mit) [reply tools/licenses/lib/licenses.dart:924]: Done

File: tools/licenses/lib/main.dart (hunk 1 above)
Comment (mit-mit) [reply tools/licenses/lib/main.dart:45]: Done

File: tools/licenses/lib/licenses.dart (hunk 5 above)
Comment (mit-mit) [reply tools/licenses/lib/licenses.dart]: I was missing a set of parenthesis :-)

```
-      while (hadMoreLines = lines.moveNext() && lines.current.value.startsWith(prefix)) {
+      while ((hadMoreLines = lines.moveNext()) && lines.current.value.startsWith(prefix)) {
```

**Comment by mit-mit** [comment]
Test is now passing, but I need to update the goldens:
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8803139310453266113/+/u/licenses_check/stdout

**Comment by mit-mit** [comment]
Golden updated:
```
mit@mit-gbuntu:~/dev/flutter_engine/src/flutter/tools/licenses$ rm -rf ../../../out/licenses
dart lib/main.dart --src ../../.. --out ../../../out/licenses --golden ../../ci/licenses_golden
Finding files...

Computing signature for license tool
    Detected changes to license tool. Forcing license collection for all components.
Collecting licenses for flutter
      2918 of 2918 ██████████ 100% (0 missing licenses)  Done.
Collecting licenses for fuchsia
      1831 of 1831 ██████████ 100% (0 missing licenses)  Done.
Collecting licenses for gpu
         2 of 2 ██████████ 100% (0 missing licenses)  Done.
Collecting licenses for third_party
     11611 of 11611 ██████████ 100% (0 missing licenses)  Done.
Collecting licenses for skia
      5296 of 5296 ██████████ 100% (0 missing licenses)  Done.
mit@mit-gbuntu:~/dev/flutter_engine/src/flutter/tools/licenses$
mit@mit-gbuntu:~/dev/flutter_engine/src/flutter/tools/licenses$ cp ../../../out/licenses/* ../../ci/licenses_golden
git diff
diff --git a/ci/licenses_golden/tool_signature b/ci/licenses_golden/tool_signature
index 74aacad982..6ae74cafe6 100644
--- a/ci/licenses_golden/tool_signature
+++ b/ci/licenses_golden/tool_signature
@@ -1,2 +1,2 @@
-Signature: d232ec47cb6d61b7385bf1ef936a93c4
+Signature: 647326c44237e255694c24deb9896c68
```

File: tools/licenses/lib/licenses.dart (hunk 5 above)
Comment (Hixie) [reply tools/licenses/lib/licenses.dart]: aha, glad i pinned down the section if not the reason :-)

File: tools/licenses/lib/main.dart (hunk 4 above)
Comment (Hixie) [reply tools/licenses/lib/main.dart]: yeah, the root case is special in various ways such that the code never gets called when parent is null.

This code was designed for a nullable world. I would write it quite differently now with null safety.

**Review by Hixie — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
This PR reminds me that I have a big patch (which is going to conflict up the wazoo now, oops! oh well) that was waiting for https://dart-review.googlesource.com/c/sdk/+/250792 to get reviewed and landed, but I was trying to follow the non-Googler path to contributing to Dart and didn't figure out how to get a review and then forgot about it.

**Comment by Hixie** [comment]
test-exempt: code refactor with no semantic change

**Pull request merged**